java-engine/src/main/java/io/getunleash/engine/YggdrasilFFI.java

```diff
@@ -0,0 +1,72 @@
+package io.getunleash.engine;
+
+import com.sun.jna.Library;
+import com.sun.jna.Native;
+import com.sun.jna.Platform;
+import com.sun.jna.Pointer;
+
+import java.nio.file.Paths;
+
+interface UnleashFFI extends Library {
+
+    Pointer new_engine();
+
+    void free_engine(Pointer ptr);
+
+    Pointer take_state(Pointer ptr, String toggles);
+
+    Pointer check_enabled(Pointer ptr, String name, String context);
+
+    Pointer check_variant(Pointer ptr, String name, String context);
+
+    void free_response(Pointer pointer);
+}
+
+class YggdrasilFFI {
+    private final UnleashFFI ffi; // instance ffi
+    private final Pointer enginePtr;
+
+    /**
+     * If we want singleton we just make the constructors private
+     */
+    YggdrasilFFI() {
+        this(System.getenv("YGGDRASIL_LIB_PATH"));
+    }
+
+    YggdrasilFFI(String libraryPath) {
+        if (libraryPath == null) {
+            libraryPath = "."; // assume it's accessible in current path
+        }
+        String libImpl = "libyggdrasilffi.so";
+        if (Platform.isMac()) {
+            libImpl = "libyggdrasilffi.dylib";
+        } else if (Platform.isWindows()) {
+            libImpl = "libyggdrasilffi.dll";
+        }
+
+        String combinedPath = Paths.get(libraryPath, libImpl).toString();
+
+        this.ffi = Native.load(combinedPath, UnleashFFI.class);
+        this.enginePtr = this.ffi.new_engine();
+    }
+
+    Pointer takeState(String toggles) {
+        return this.ffi.take_state(this.enginePtr, toggles);
+    }
+
+    void free() {
```

Comment:
This little beast has to get called when this falls out of scope, otherwise we end up with a memory leak. ChatGPT says a Cleaner class is the right way to go here but I'm not very familiar with that.

Reply:
Yes, I wanted to solve that in a separate PR because that's already how it is